### PR TITLE
snapcast 0.26.0 (new formula)

### PR DESCRIPTION
**Pull request by codetheweb**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

File: Formula/snapcast.rb [hunk 1]
```diff
@@ -0,0 +1,35 @@
+class Snapcast < Formula
+  desc "Synchronous multiroom audio player"
+  homepage "https://github.com/badaix/snapcast"
+  url "https://github.com/badaix/snapcast/archive/refs/tags/v0.26.0.tar.gz"
+  sha256 "166353267a5c461a3a0e7cbd05d78c4bfdaebeda078801df3b76820b54f27683"
+  license "GPL-3.0-only"
+
+  depends_on "boost" => :build
+  depends_on "cmake" => :build
```
Comment (carlocab) [inline Formula/snapcast.rb:9]: You also need `pkg-config`.

File: Formula/snapcast.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  depends_on "opus"
+  depends_on "pulseaudio"
+  uses_from_macos "expat"
+
+  def install
+    system "cmake", "-S", ".", "-B", "build", *std_cmake_args
+    system "cmake", "--build", "build"
+    system "cmake", "--install", "build"
+    # Fix permissions
+    chmod(0555, "#{share}/snapserver/plug-ins/meta_mpd.py")
```
Comment (carlocab) [inline Formula/snapcast.rb:22]: What permissions did this have originally, and why are they wrong? Shouldn't CMake handle this?

File: Formula/snapcast.rb [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  end
+
+  test do
+    pid = fork do
+      exec ("#{bin}/snapserver")
+    end
+    sleep 2
+    output = shell_output("timeout 2 #{bin}/snapclient", 124)
```
Comment (carlocab) [inline Formula/snapcast.rb:30]: I don't think relying on `timeout` is portable.

File: Formula/snapcast.rb (hunk 1 above)
Comment (codetheweb) [reply Formula/snapcast.rb:9]: Fixed in [529f17a](https://github.com/Homebrew/homebrew-core/pull/107089/commits/529f17a71e04f4c437904a716d72f0c9bd27c06f).

**Comment by SMillerDev** [comment]
```
Errno::ENOENT: No such file or directory - timeout
```

File: Formula/snapcast.rb (hunk 3 above)
Comment (codetheweb) [reply Formula/snapcast.rb]: Fixed in 50317d72280.

File: Formula/snapcast.rb (hunk 2 above)
Comment (codetheweb) [reply Formula/snapcast.rb:23]: The original permissions were `111`. I'm not sure why they're wrong, but if nothing is changed Homebrew throws with

```
Error: Failed to read Mach-O binary: /opt/homebrew/Cellar/snapcast/0.26.0/share/snapserver/plug-ins/meta_mpd.py
```

Comment (carlocab) [reply Formula/snapcast.rb:23]: Can you add that to your comment? Preferably with a `FIXME`.

File: Formula/snapcast.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+      stdout.each { |line| output << line }
+      rescue EOFError
+        break
+    end
+
+    begin
+      timeout(2) { Process.wait(client_pid) }
+    rescue Timeout::Error
+      break
+    end
```
Comment (carlocab) [inline Formula/snapcast.rb:42]: Why not just
```suggestion
    sleep 2
    Process.kill "TERM", client_pid
```
?

Or something like that. Or if we don't want to kill the client, just `sleep`.

**Comment by codetheweb** [comment]
It looks like the build is failing on Ubuntu 16 because the source has 
```
#include <optional>
```
which requires C++17. Is there an easy way to require a more recent version of GCC?

File: Formula/snapcast.rb (hunk 2 above)
Comment (codetheweb) [reply Formula/snapcast.rb:28]: Fixed in 67e1739eb40.

File: Formula/snapcast.rb (hunk 4 above)
Comment (codetheweb) [reply Formula/snapcast.rb]: Yep, refactored in 85031132f7e.`

**Comment by carlocab** [comment]
That'll be fixed when we start building with a newer Linux distro verrsion, which I expect to be done in the next few weeks or so. (But, it's open source, so this timeline could be totally wrong.) See https://github.com/Homebrew/brew/issues/13619.

In the meantime, I suggest we ignore the Linux failure and just make sure this builds on macOS.

File: Formula/snapcast.rb [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  depends_on "boost" => :build
+  depends_on "cmake" => :build
+  depends_on "pkg-config" => :build
+  depends_on "flac"
+  depends_on "libsoxr"
+  depends_on "libvorbis"
+  depends_on "opus"
+  depends_on "pulseaudio"
+  uses_from_macos "expat"
+  on_linux do
```
Comment (carlocab) [inline Formula/snapcast.rb:17]:
```suggestion
  uses_from_macos "expat"

  on_linux do
```

File: Formula/snapcast.rb [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    depends_on "avahi"
+  end
+
+  def install
+    system "cmake", "-S", ".", "-B", "build", *std_cmake_args
+    system "cmake", "--build", "build"
+    system "cmake", "--install", "build"
+    # FIXME: if permissions aren't changed, the install fails with:
+    # Error: Failed to read Mach-O binary: share/snapserver/plug-ins/meta_mpd.py
+    chmod(0555, "#{share}/snapserver/plug-ins/meta_mpd.py")
```
Comment (carlocab) [inline Formula/snapcast.rb:28]:
```suggestion
    chmod 0555, share/"snapserver/plug-ins/meta_mpd.py"
```

File: Formula/snapcast.rb [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+  end
+
+  test do
+    server_pid = fork do
+      exec ("#{bin}/snapserver")
```
Comment (carlocab) [inline Formula/snapcast.rb:33]:
```suggestion
      exec bin/"snapserver"
```

File: Formula/snapcast.rb [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    end
+
+    r, w = IO.pipe
+    client_pid = spawn("#{bin}/snapclient", out: w)
```
Comment (carlocab) [inline Formula/snapcast.rb:37]:
```suggestion
    client_pid = spawn bin/"snapclient", out: w
```

File: Formula/snapcast.rb [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    sleep 5
+    Process.kill("SIGTERM", client_pid)
+
+    output = r.read
+    r.close
+
+    assert_match(/Connected to/m, output)
+  ensure
+    Process.kill("SIGTERM", server_pid)
+    Process.kill("SIGTERM", client_pid)
```
Comment (carlocab) [inline Formula/snapcast.rb:49]: This seems redundant.

File: Formula/snapcast.rb [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+    client_pid = spawn("#{bin}/snapclient", out: w)
+    w.close
+
+    sleep 5
+    Process.kill("SIGTERM", client_pid)
+
+    output = r.read
+    r.close
+
+    assert_match(/Connected to/m, output)
```
Comment (carlocab) [inline Formula/snapcast.rb:46]:
```suggestion
    assert_match("Connected to", output)
```
Why did we need the `m`?

**Review by carlocab — COMMENTED** [review]
Just some minor style suggestions then I'll be happy to merge this.

File: Formula/snapcast.rb (hunk 10 above)
Comment (codetheweb) [reply Formula/snapcast.rb]: I thought `assert_match` was doing a stricter check but yep that works. :)

**Comment by codetheweb** [comment]
Thanks for all the comments!

**Review by carlocab — APPROVED** [review]
Thanks!

**Pull request closed without merging**